# Problem 1

First, we review the following theorems:

Intermediate Value Theorem:

Let the function $f$ be defined on the interval $I = [c,d]$. Let $M$ be the maximum of $f$ on $[c,d]$ and let $m$ be the minimum of $f$ on $[c,d]$. If $y$ is any real number between $m$ and $M$, then there exists some real number $x \in [c,d]$ such that $f(x) = y$.

Mean Value Theorem:
Let the function $f$ be continuous on the closed interval $[c,d]$ and differentiable on the open interval $(c,d)$. Then, there exists some point $x \in (c,d)$ such that $$f^\prime(x) = \frac{f(d)-f(c)}{d-c}$$

Rolle's Theorem:
Let the function $f$ be continuous on the closed interval $[c,d]$ and differentiable on the open interval $(c,d)$. Furthermore, suppose that $f(c) = f(d)$. Then, there exists some point $x \in (c,d)$ such that $$f^\prime(x) = 0$$

Mean Value Theorem for Integrals:
Let the function $f$ be continuous on the closed interval $[c,d]$. Then, there exists some point $a \in [c,d]$ such that $$f(a) = \frac{1}{d-c}\int_c^d f(x) dx$$

Weighted Mean Value Theorem for Integrals:
Let the functions $f$ and $g$ be continuous on the closed interval $[c,d]$. If $g(x) \geq 0$ or $g(x) \leq 0$ for all $x \in [c,d]$, then there exists some point $a \in [c,d]$ such that $$ \int_c^d f(x)g(x) dx = f(a) \int_c^d g(x) dx$$

# Problem 2

In [62]:
#This code implements the composite Trapezoidal rule quadrature. Written by Ethan Martirosyan. Last edited on 10/2/2024.
#First, we import the function exp from the math module in order to define the function that we will be approximating
from math import exp

#The following function actually computes the trapezoid rule.
#It takes the inputs a (starting value), b (stopping value), N (number of steps), and f
#It outputs the trapezoidal approximation of the function
def trapezoid_rule(a,b,N,f):
    #We initialize the ans variable
    ans = 0
    #Now, we define the step size
    h = (b-a)/N
    #Here we take the sum of f(x_1) + ... + f(x_n-1) and add it to the ans variable 
    for x in range(1,N):
        ans += f(a + h * x)
    #Now we add (1/2)f(a) and (1/2)f(b) to the sum. Notice that it would be unwieldy to do this in the for loop.    
    ans += (1/2)*f(a)
    ans += (1/2)*f(b)
    #Now we multiply by h to obtain the final answer
    ans = h * ans
    #Return the final answer
    return ans



# Problem 3

In [75]:
#Here we define the function in the problem statement
def func(val):
    return (val) * exp((val)**2)

#We print the approximations for various values of N
#Print the approximation for N = 10 (or h = 1/10)
print("10 steps:", trapezoid_rule(0,1,10,func))
#Print the approximation for N=20 (or h = 1/20)
print("20 steps:", trapezoid_rule(0,1,20,func))
#Print the approximation for N = 40 (or h = 1/40)
print("40 steps:", trapezoid_rule(0,1,40,func))

10 steps: 0.8650898321786331
20 steps: 0.8606306634063897
40 steps: 0.8595135096608248


First, we note that $\int_0^1 xe^{x^2} dx = \frac{e-1}{2} \approx 0.85914091422952$.

Let us compare the difference between our approximation $T_h[f]$ and the actual value $I[f]$ for different values of $h$. 

For $h = 1/10$, we have $T_{1/10}[f] = 0.8650898321786331$. 

For $h = 1/20$, we have $T_{1/20}[f] = 0.8606306634063897$. 

For $h = 1/40$, we have $T_{1/40}[f] = 0.8595135096608248$.

Now, we compute the differences: 
$$\vert I[f] - T_{1/10}[f] \vert = \vert 0.85914091422952 - 0.8650898321786331 \vert = 0.0059489179491131$$
$$\vert I[f] - T_{1/20}[f] \vert = \vert 0.85914091422952 - 0.8606306634063897 \vert = 0.0014897491768697$$
$$\vert I[f] - T_{1/40}[f] \vert = \vert 0.85914091422952 - 0.8595135096608248\vert =  0.0003725954313048$$
Notice that
$$ \frac{\vert I[f] - T_{1/10}[f] \vert}{\vert I[f] - T_{1/20}[f] \vert} = \frac{0.0059489179491131}{0.0014897491768697} \approx 3.993234 \approx 4$$
$$ \frac{\vert I[f] - T_{1/20}[f] \vert}{\vert I[f] - T_{1/40}[f] \vert} = \frac{0.0014897491768697}{0.0003725954313048} \approx 3.9983 \approx 4$$ This confirms the quadratic rate of decay of the error.

# Problem 4

In [66]:
from math import exp
#This is the function that we are trying to approximate for problem 4
def func2(val):
    return exp(-1*(val**2))
#This computes the trapezoidal rule again except we use h instead of N this time
def trapezoid_rule_2(a,b,h,f):
    #define the final answer that will be returned
    ans = 0
    #add (1/2)f(a) and (1/2)f(b) to this answer
    ans = ans + (1/2)*f(a) + (1/2)*f(b)
    #define the counter that will be used to sum f(x_1) + ... + f(x_n-1)
    count = 1
    #sums up f(x_1) + ... + f(x_n-1)
    while a+count*h < b:
        ans = ans + f(a+count*h)
        count+=1
    #make sure to mulitply the answer by h
    ans = ans * h
    #return the answer
    return ans
#computes the ratio q(h)
def q(h):
    return (trapezoid_rule_2(0,1,h/2,func2) - trapezoid_rule_2(0,1,h,func2))/(trapezoid_rule_2(0,1,h/4,func2) - trapezoid_rule_2(0,1,h/2,func2))
#Now, we wish to find a value of h such that q(h) is close to 4
#To start, set the value of h to 1
h = 1
#Keep dividing h by 2 until q(h) is close to 4
while abs(q(h) - 4) > 0.0000001:
    h /= 2
#Now we print out this value of h
print("The desired value of h is", h)
#We also print out the corresponding value of q(h)
print("At this value of h, we find that q(h) is", q(h))


The desired value of h is 0.0009765625
At this value of h, we find that q(h) is 3.9999999088625047


Now, we will obtain an approximation of the error $I[e^{-x^2}] - T_h[e^{-x^2}]$. 
According to the notes from class, we know that a good approximation of the error is as follows:
$$\frac{4}{3} [T_{h/2}[f] - T_h[f]]$$

In [68]:
#Compute the error
error = (4/3)*(trapezoid_rule_2(0,1,h/2,func2)-trapezoid_rule_2(0,1,h,func2))
#Print out this error
print("The approximate error is", error)
#Compute the approximate answer by adding the error to our estimate
approx_answer = trapezoid_rule_2(0,1,h,func2) + error
print("The approximate value of the integral is", approx_answer)

The approximate error is 5.847286432825172e-08
The approximate value of the integral is 0.7468241328124272


$S_h(f)$

The reason why $$S_h[e^{-x^2}] = T_h[e^{-x^2}] + \frac{4}{3} (T_{h/2}[e^{-x^2}] - T_h[e^{-x^2}])$$ is a better approximation than $$T_h[e^{-x^2}]$$ is because we are adding the approximate error in the form of $$\frac{4}{3} (T_{h/2}[e^{-x^2}] - T_h[e^{-x^2}])$$ Furthermore, we know that $S_h[e^{-x^2}]$ converges to the actual value of the integral faster than $T_h[e^{-x^2}]$ because $$I[f] - T_h[f]$$ is $O(h^2)$ whereas $$I[f] - S_h[f]$$ is $o(h^2)$ (or $O(h^4)$)).